relay-replays/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+//! Relay processing and normalization module.
+//!
+//! Replays are multi-part values sent from Sentry integrations spanning arbitrary time-periods.
+//! They are ingested incrementally and are available for viewing after the first two segments
+//! (segment 0 and 1) have completed ingestion.
+//!
+//! # Protocol
+//!
+//! Relay is expecting a JSON object with some mandatory metadata.  However, environment and user
+//! metadata is usually sent in addition to the minimal payload.
+//!
+//! ```json
+//! {
+//!     "type": "replay_event",
+//!     "replay_id": "d2132d31b39445f1938d7e21b6bf0ec4",
+//!     "event_id": "63c5b0f895441a94340183c5f1e74cd4",
+//!     "segment_id": 0,
+//!     "timestamp": 1597976392.6542819,
+//!     "replay_start_timestamp": 1597976392.6542819,
+//!     "urls": ["https://sentry.io"],
+//!     "error_ids": ["d2132d31b39445f1938d7e21b6bf0ec4"],
+//!     "trace_ids": ["63c5b0f895441a94340183c5f1e74cd4"],
+//!     "requests": {
+//!         "headers": {"User-Agent": "Mozilla/5.0..."}
+//!     },
+//! }
+//! ```
+use relay_general::user_agent::{parse_device, parse_os, parse_user_agent, Device};
+use serde::{Deserialize, Serialize};
+use serde_json::Error;
+use std::collections::HashMap;
+use std::fmt::Write;
+use std::net::IpAddr;
+
+pub fn normalize_replay_event(
+    replay_bytes: &[u8],
+    detected_ip_address: Option<IpAddr>,
+) -> Result<Vec<u8>, Error> {
+    let mut replay_input: ReplayInput = serde_json::from_slice(replay_bytes)?;
+
+    // Set user-agent metadata.
+    replay_input.set_user_agent_meta();
+
+    // Set user ip-address if needed.
+    match detected_ip_address {
+        Some(ip_address) => replay_input.set_user_ip_address(ip_address),
+        None => (),
+    }
+
+    serde_json::to_vec(&replay_input)
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct ReplayInput {
```

Comment:
This structure is public but its fields are not. Should the fields be made public?

Reply:
Made the struct private.  I don't see why it should be public unless you know better.